### PR TITLE
Fixed a bug that caused a syntax error when a function was specified to server_default when creating a column in MySQL.

### LINKED ISSUE #11317: Generated syntax error when using funcs in server_default in MySQL
### Discussed in https://github.com/sqlalchemy/sqlalchemy/discussions/11315

<div type='discussions-op-text'>

<sup>Originally posted by **martinm07** April 24, 2024</sup>
Using MySQL version `8.0.29` and SQLAlchemy version `2.0.29` on Windows 10, the following code produces the further following error:
```python
from sqlalchemy import create_engine, Table, Column, MetaData, Integer, Date, func

engine = create_engine("mysql://root:password@localhost/dbname", echo=True)

metadata_obj = MetaData()
user_table = Table(
    "users",
    metadata_obj,
    Column("id", Integer, primary_key=True),
    Column("date_created", Date, server_default=func.current_date())
)

metadata_obj.create_all(engine)
```
```
sqlalchemy.exc.ProgrammingError: (MySQLdb.ProgrammingError) (1064, "You have an error in your SQL syntax; check the manual that corresponds 
to your MySQL server version for the right syntax to use near 'CURRENT_TIME, \n\tPRIMARY KEY (id)\n)' at line 3")
[SQL:
CREATE TABLE users (
        id INTEGER NOT NULL AUTO_INCREMENT,
        date_created DATE DEFAULT CURRENT_TIME,
        PRIMARY KEY (id)
)

]
(Background on this error at: https://sqlalche.me/e/20/f405)
```
SQLAlchemy produces incorrect MySQL code, which should be what I put in `text()` below:
```python
from sqlalchemy import create_engine, text

engine = create_engine("mysql://root:WeMustKnowWeWillKnow@localhost/testing", echo=True)

with engine.connect() as conn:
    conn.execute(
        text("""
CREATE TABLE users (
        id INTEGER NOT NULL AUTO_INCREMENT,
        date_created DATE DEFAULT (CURRENT_TIME),
        PRIMARY KEY (id)
)
""")
    )
    conn.commit()
```
Namely, brackets around `CURRENT_TIME`.
Is this a bug, or some known limitation?</div>

### PR DESCRIPTION
**Pull request by huuyafwww**
Fixes #11317 

### Description
Fix a syntax error bug by inserting parentheses when a function is specified in the process of setting column default values during table creation within MySQLDDLCompiler.
I also added the corresponding test cases.

### Checklist

This pull request is:

- [ ] A documentation / typographical / small typing error fix
	- Good to go, no issue or tests are needed
- [x] A short code fix
	- issue is #11317 
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.   one line code fixes without tests will not be accepted.
- [ ] A new feature implementation
	- please include the issue number, and create an issue if none exists, which must
	  include a complete example of how the feature would look.
	- Please include: `Fixes: #<issue number>` in the commit message
	- please include tests.

**Have a nice day!**

### CODE REVIEW

**Review by huuyafwww — COMMENTED** [review]
Hello.
Thank you for developing a great library.
I am having trouble with systax error when setting a function in server_default when creating a table in MySQL as per the pull request title.
Would you accept this fix?

File: lib/sqlalchemy/dialects/mysql/base.py [hunk 1]
```diff
@@ -1853,6 +1853,11 @@ def get_column_specification(self, column, **kw):
                 colspec.append("DEFAULT " + default)
         return " ".join(colspec)
 
+    def get_column_default_string(self, column):
+        if hasattr(column.server_default, 'arg') and isinstance(column.server_default.arg, functions.FunctionElement):
```
Comment (CaselIT) [inline lib/sqlalchemy/dialects/mysql/base.py:1857]: can you change add the parents in get_column_specification similar to how it's done in sqlite? https://github.com/sqlalchemy/sqlalchemy/blob/fe03298593e59e5d40de81e7f864690e73930666/lib/sqlalchemy/dialects/sqlite/base.py#L1548

**Review by CaselIT — CHANGES_REQUESTED** [review]
Thanks, nice work!

Could you also add a changelog here? https://github.com/sqlalchemy/sqlalchemy/tree/main/doc/build/changelog/unreleased_20 (the number is the issue id)

File: lib/sqlalchemy/dialects/mysql/base.py (hunk 1 above)
Comment (huuyafwww) [reply lib/sqlalchemy/dialects/mysql/base.py]: @CaselIT 
Thank you for your prompt confirmation.
I have made the corrections, are they in line with the suggested corrections you have indicated?

https://github.com/sqlalchemy/sqlalchemy/pull/11953/commits/9b0da3cce385c0946726c6bce7163f0eef66772c

Comment (CaselIT) [reply lib/sqlalchemy/dialects/mysql/base.py]: works for me, can add a changelog? In the meantime I've made the ci run to see if there are style issues

Comment (huuyafwww) [reply lib/sqlalchemy/dialects/mysql/base.py]: @CaselIT 
Thank you very much.
Change history has been added.
https://github.com/sqlalchemy/sqlalchemy/pull/11953/commits/cc07152f66a49c929563206aa849fac173eed52d

Comment (huuyafwww) [reply lib/sqlalchemy/dialects/mysql/base.py]: Fixed a lint error.
https://github.com/sqlalchemy/sqlalchemy/pull/11953/commits/ea4824fff993cd2174359f370cb7358f45b708f5

File: doc/build/changelog/unreleased_20/11317.rst [hunk 2]
```diff
@@ -0,0 +1,6 @@
+.. change::
+    :tags: bug, orm
```
Comment (CaselIT) [inline doc/build/changelog/unreleased_20/11317.rst:2]:
```suggestion
    :tags: bug, schema
```

**Comment by CaselIT** [comment]
Great, thanks. Let's run the full CI.

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision c67ac6868f111c8045c64a889386662446f666d8 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
New Gerrit review created for change c67ac6868f111c8045c64a889386662446f666d8: https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5515

**Review by CaselIT — APPROVED** [review]

**Comment by CaselIT** [comment]
oh great. they broke compat with 5.7 here. 
>  (1064, "You have an error in your SQL syntax; check the manual that corresponds to your MySQL server version for the right syntax to use near '(CURRENT_TIMESTAMP)\n)' at line 2")
[SQL: 
CREATE TABLE hasts (
	x DATETIME NOT NULL DEFAULT (CURRENT_TIMESTAMP)
)


really great idea here by mysql...

We need to gate this on server version. Do we know if the change has been in 8.0.0 or at some other point?

**Comment by huuyafwww** [comment]
@CaselIT 
Thanks for the report on the test results.

> We need to gate this on server version. Do we know if the change has been in 8.0.0 or at some other point?

I have done some research and it seems to be supported by [8.0.13](https://dev.mysql.com/doc/refman/8.0/en/data-type-defaults.html) for MySQL and [10.2.1](https://mariadb.com/kb/en/mariadb-1021-release-notes/) or later for MariaDB.
Corresponding corrections were made in the following commits
https://github.com/sqlalchemy/sqlalchemy/pull/11953/commits/d93ac419a9201134e9c4845dd2e4dc48db4b6f78

**Comment by CaselIT** [comment]
Thanks!

**Review by sqla-tester — COMMENTED** [review]
OK, this is **sqla-tester** setting up my work on behalf of **CaselIT** to try to get revision d93ac419a9201134e9c4845dd2e4dc48db4b6f78 of this pull request into gerrit so we can run tests and reviews and stuff

**Comment by sqla-tester** [comment]
Patchset d93ac419a9201134e9c4845dd2e4dc48db4b6f78 added to existing Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5515

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5515 has been **merged**. Congratulations! :)

**Comment by sqla-tester** [comment]
Gerrit review https://gerrit.sqlalchemy.org/c/sqlalchemy/sqlalchemy/+/5516 has been **merged**. Congratulations! :)

**Pull request closed without merging**